[cocoa] Begin adding https navigation failure browsing warning

sysrqb opened the pull request:
#### 743aecc8bcfee0db00fd09778a5becc4cf46d53a
<pre>
[cocoa] Begin adding https navigation failure browsing warning
<a href="https://bugs.webkit.org/show_bug.cgi?id=277527">https://bugs.webkit.org/show_bug.cgi?id=277527</a>
<a href="https://rdar.apple.com/problem/133035983">rdar://problem/133035983</a>

Reviewed by Alex Christensen.

Add two new types that identify the desired warning. Currently the supported
warnings are for SafeBrowsing and HTTPSNavigationFailure.
HTTPSNavigationFailure is the warning associated with the
https-by-default/HTTPSOnly load failure.

The current HTTPSNavigationFailure warning uses the same buttons, color, and
image as the SafeBrowsing warning. These will be changed in subsequent patches.

The SafeBrowsing warning is covered by existing tests. The
HTTPSNavigationFailure warning will be covered by tests in a later patch.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showWarningViewWithURL:title:warning:detailsWithLinks:completionHandler:]):
* Source/WebKit/UIProcess/BrowsingWarning.h:
(WebKit::BrowsingWarning::create):
(WebKit::BrowsingWarning::type const):
* Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
(WebKit::browsingWarningTitleText):
(WebKit::browsingWarningText):
(WebKit::browsingDetailsText):
(WebKit::BrowsingWarning::BrowsingWarning):
(WebKit::browsingTitleText): Deleted.
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(backgroundColorForWarningType):
(-[_WKWarningView initWithFrame:browsingWarning:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/281773@main">https://commits.webkit.org/281773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b26a18a1f2f98983d9a6d8c082ade987d3855ea4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13514 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64870 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11485 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11760 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49269 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7975 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62972 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/37516 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52798 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30094 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34199 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10017 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10398 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/56026 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10313 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66600 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4882 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10141 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56635 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4904 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56825 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4044 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9174 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37184 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38277 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
```diff
@@ -260,6 +260,17 @@ static CGSize buttonSize(ButtonType *button)
 #endif
 }
 
+static WarningItem backgroundColorForWarningType(const WebKit::BrowsingWarning& warning)
```
This doesn't make any sense.  I would expect this to return a color, but instead it returns WarningItem::SafeBrowsingBackground even if that isn't accurate.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingWarning.h:
```diff
@@ -78,6 +86,7 @@ class BrowsingWarning : public RefCounted<BrowsingWarning> {
 #if PLATFORM(COCOA)
     RetainPtr<NSAttributedString> m_details;
 #endif
+    WarningType m_type;
```
const WarningType?

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
```diff
@@ -116,6 +116,18 @@ static void addLinkAndReplace(NSMutableAttributedString *string, NSString *toRep
     return WEB_UI_NSSTRING(@"Website With Harmful Software Warning", "Unwanted software warning title");
 }
 
+static NSString *browsingWarningTitleText(BrowsingWarning::WarningType type)
```
This isn't a great abstraction.  It takes a type, but seems like it should only be called with WarningType::HTTPSNavigationFailure.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
```diff
@@ -127,6 +139,18 @@ static void addLinkAndReplace(NSMutableAttributedString *string, NSString *toRep
     return WEB_UI_NSSTRING(@"This website may try to trick you into installing software that harms your browsing experience, like changing your settings without your permission or showing you unwanted ads. Once installed, it may be difficult to remove.", "Unwanted software warning");
 }
 
+static NSString *browsingWarningText(BrowsingWarning::WarningType type)
```
Ditto.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
```diff
@@ -164,21 +188,45 @@ static void addLinkAndReplace(NSMutableAttributedString *string, NSString *toRep
 {
 }
+
+BrowsingWarning::BrowsingWarning(const URL& url, bool forMainFrameNavigation, WarningType type)
```
This seems like it would be bad to be called with WarningType::SafeBrowsing.  Instead of two constructors, I think we need something like this:

struct HTTPSNavigationFailureData { };
struct SafeBrowsingWarningData {
    RetainPtr<SSBServiceLookupResult> result;
};
using Data = std::variant<SafeBrowsingWarningData, HTTPSNavigationFailureData>;
BrowsingWarning(const URL& url, bool forMainFrameNavigation, Data&&);

Then use that variant instead of these strange functions only intended for one type.

Commit:
[cocoa] Begin adding https navigation failure browsing warning
https://bugs.webkit.org/show_bug.cgi?id=277527
rdar://problem/133035983

Reviewed by Alex Christensen.

Add two new types that identify the desired warning. Currently the supported
warnings are for SafeBrowsing and HTTPSNavigationFailure.
HTTPSNavigationFailure is the warning associated with the
https-by-default/HTTPSOnly load failure.

The current HTTPSNavigationFailure warning uses the same buttons, color, and
image as the SafeBrowsing warning. These will be changed in subsequent patches.

The SafeBrowsing warning is covered by existing tests. The
HTTPSNavigationFailure warning will be covered by tests in a later patch.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showWarningViewWithURL:title:warning:detailsWithLinks:completionHandler:]):
* Source/WebKit/UIProcess/BrowsingWarning.h:
(WebKit::BrowsingWarning::create):
(WebKit::BrowsingWarning::type const):
* Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
(WebKit::browsingWarningTitleText):
(WebKit::browsingWarningText):
(WebKit::browsingDetailsText):
(WebKit::BrowsingWarning::BrowsingWarning):
(WebKit::browsingTitleText): Deleted.
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(backgroundColorForWarningType):
(-[_WKWarningView initWithFrame:browsingWarning:completionHandler:]):

Canonical link: https://commits.webkit.org/281773@main
